<a href="https://colab.research.google.com/github/de-fernandez/CPE-313-CPE32S3/blob/main/Don%20Eleazar%20T.%20Fernandez%2C%20CPE32S3%2C%20CPE%20313%20(Hands-on%20Activity%208.1%20-%20Text%20Similarity).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Activity 5.1 Text Similarity

#### Objective(s):

This activity aims to introduce how to implement text similarity

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to remove special characters from text
* Demonstrate how to measure text simiarity using distance
* Demonstrate how to find text similarity using Count Vector and TF-IDF Vectorizer


#### Resources:
* Jupyter Notebook
* songdata.csv

## Procedures
Load the necessary libraries

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import nltk
import pandas as pd

In [3]:
data = pd.read_csv('/content/drive/MyDrive/CPE 313/H.O.A. 8.1/songdata.csv')
data.head()

,artist,song,link,text
0,ABBA,Ahe's My Kind Of Girl,/a/abba/ahes+my+kind+of+girl_20598417.html,"Look at her face, it's a wonderful face \nAnd..."
1,ABBA,"Andante, Andante",/a/abba/andante+andante_20002708.html,"Take it easy with me, please \nTouch me gentl..."
2,ABBA,As Good As New,/a/abba/as+good+as+new_20003033.html,I'll never know why I had to go \nWhy I had t...
3,ABBA,Bang,/a/abba/bang_20598415.html,Making somebody happy is a question of give an...
4,ABBA,Bang-A-Boomerang,/a/abba/bang+a+boomerang_20002668.html,Making somebody happy is a question of give an...


Find the lyrics by the artist The Beatles

In [4]:
beatles = data[data.artist=='The Beatles']
beatles.head()

,artist,song,link,text
1198,The Beatles,A Shot Of Rhythm And Blues,/b/beatles/a+shot+of+rhythm+blues_20014867.html,"Well, if your hands start a-clappin' \nAnd yo..."
1199,The Beatles,Across The Universe,/b/beatles/across+the+universe_10026507.html,Words are flowing out like \nEndless rain int...
1200,The Beatles,All I've Got To Do,/b/beatles/all+ive+got+to+do_10026646.html,"Whenever I want you around, yeah \nAll I gott..."
1201,The Beatles,And I Love Her,/b/beatles/and+i+love+her_10026463.html,I give her all my love \nThat's all I do \nA...
1202,The Beatles,And Your Bird Can Sing,/b/beatles/and+your+bird+can+sing_10026364.html,You tell me that you've got everything you wan...


Check the number of Beatles songs in the data set


In [5]:
data.artist.value_counts().loc[['The Beatles']]

,count
artist,
The Beatles,178


Take a look at the first song's lyrics

In [6]:
beatles.iloc[0,3]

"Well, if your hands start a-clappin'  \nAnd your fingers start a-poppin'  \nAnd your feet start a-movin' around  \nAnd if you start to swing and sway  \n  \nWhen the band starts to play  \nA real cool way out sound  \nAnd if you get to can't help it and you can't sit down  \nYou feel like you gotta move around  \n  \nYou get a shot of rhythm and blues.  \nWith just a little rock and roll on the side  \nJust for good measure.  \nGet a pair of dancin' shoes  \n  \nWell, with your lover by your side  \nDon't you know you're gonna have a rockin' time, see'mon!  \nDon't you worry 'bout a thing  \nIf you start to dance and sing  \n  \nAnd chills come up on you  \nAnd if the rhythm finally gets you and the beat gets you too  \nWell, here's something for you to do  \n  \nGet a shot of rhythm and blues  \nWith just a little rock and roll on the side  \nJust for good measure  \nGet a pair of dancin' shoes  \n  \nWell, with your lover by your side  \nDon't you know you're gonna have a rockin' ti

Remove characters from the lyrics

In [7]:
import re

newline = lambda x: re.sub('\n', ' ', x) # remove \n
alphanumeric = lambda x: re.sub('\w*\d\w*', ' ', x) # remove alphanumeric words

corpus = beatles.text.map(newline).map(alphanumeric)
corpus.head()

<>:4: SyntaxWarning: invalid escape sequence '\w'
<>:4: SyntaxWarning: invalid escape sequence '\w'
/tmp/ipykernel_31196/2279837142.py:4: SyntaxWarning: invalid escape sequence '\w'
  alphanumeric = lambda x: re.sub('\w*\d\w*', ' ', x) # remove alphanumeric words


,text
1198,"Well, if your hands start a-clappin' And you..."
1199,Words are flowing out like Endless rain into...
1200,"Whenever I want you around, yeah All I gotta..."
1201,I give her all my love That's all I do And...
1202,You tell me that you've got everything you wan...


Create a document term matrix using Count Vectorizer with the stop words turned on to English

In [8]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(stop_words="english")
X = cv.fit_transform(corpus).toarray()

dt = pd.DataFrame(X, columns=cv.get_feature_names_out()).set_index(beatles.song)
dt.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Across The Universe,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
All I've Got To Do,0,0,0,0,0,0,0,0,0,0,...,2,0,0,0,0,0,0,0,0,0
And I Love Her,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
And Your Bird Can Sing,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


Set display options so that the lyrics aren't cut off

In [9]:
pd.set_option('display.max_colwidth', None)

In [10]:
beatles[beatles.song=='Imagine'].text

,text
24783,Imagine there's no heaven \nIt's easy if you try \nNo hell below us \nAbove us only sky \nImagine all the people \nLiving for today... \n \nImagine there's no countries \nIt isn't hard to do \nNothing to kill or die for \nAnd no religion too \nImagine all the people \nLiving life in peace... \n \nYou may say I'm a dreamer \nBut I'm not the only one \nI hope someday you'll join us \nAnd the world will be as one \n \nImagine no possessions \nI wonder if you can \nNo need for greed or hunger \nA brotherhood of man \nImagine all the people \nSharing all the world... \n \nYou may say I'm a dreamer \nBut I'm not the only one \nI hope someday you'll join us \nAnd the world will live as one\n\n


Imagine lyrics in Count Vectorizer form


In [11]:
imagine = list(dt.loc['Imagine'])
imagine[:20]

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

Define the cosine similarity calculation

In [12]:
from numpy import dot
from numpy.linalg import norm

cosine = lambda v1, v2: dot(v1, v2) / (norm(v1) * norm(v2))

Calculate all similarities and sort by the most similar

In [13]:
similarity = [cosine(imagine, song) for song in X]
sorted(list(zip(similarity, beatles.song)), reverse=True)[1:6]

[(np.float64(0.22757944185316942), "I'll Cry Instead"),
 (np.float64(0.1738270040532526), 'In My Life'),
 (np.float64(0.16574838603294895), 'Eleanor Rigby'),
 (np.float64(0.1490280831849844), 'In Spite Of All The Danger'),
 (np.float64(0.14592261028699943), 'All My Loving')]

Create the TF-IDF Vectorizer

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer

cv_tfidf = TfidfVectorizer(stop_words="english")
X_tfidf = cv_tfidf.fit_transform(corpus).toarray()

dt_tfidf = pd.DataFrame(X_tfidf, columns=cv_tfidf.get_feature_names_out()).set_index(beatles.song)
dt_tfidf.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Across The Universe,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
All I've Got To Do,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.120483,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And I Love Her,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And Your Bird Can Sing,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

cv_tfidf = TfidfVectorizer(stop_words="english")
X_tfidf = cv_tfidf.fit_transform(corpus).toarray()

dt_tfidf = pd.DataFrame(X_tfidf, columns=cv_tfidf.get_feature_names_out()).set_index(beatles.song)
dt_tfidf.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Across The Universe,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
All I've Got To Do,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.120483,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And I Love Her,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And Your Bird Can Sing,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Calculate all similarities and sort by the most similar using the TF-IDF data

In [16]:
similarity_tfidf = [cosine(imagine, song) for song in X_tfidf]
sorted(list(zip(similarity_tfidf, beatles.song)), reverse=True)[1:6]

[(np.float64(0.19266277925834221), "I'll Cry Instead"),
 (np.float64(0.18004423848592196), "I'll Get You"),
 (np.float64(0.1551371744116933), 'Eleanor Rigby'),
 (np.float64(0.15054395512050195), 'In My Life'),
 (np.float64(0.12306875209548622), 'Love Of The Loved')]

Compare the lyrics of Imagine and I'll Cry Instead

In [17]:
# Imagine lyrics
beatles[beatles.song=='Imagine'].text

,text
24783,Imagine there's no heaven \nIt's easy if you try \nNo hell below us \nAbove us only sky \nImagine all the people \nLiving for today... \n \nImagine there's no countries \nIt isn't hard to do \nNothing to kill or die for \nAnd no religion too \nImagine all the people \nLiving life in peace... \n \nYou may say I'm a dreamer \nBut I'm not the only one \nI hope someday you'll join us \nAnd the world will be as one \n \nImagine no possessions \nI wonder if you can \nNo need for greed or hunger \nA brotherhood of man \nImagine all the people \nSharing all the world... \n \nYou may say I'm a dreamer \nBut I'm not the only one \nI hope someday you'll join us \nAnd the world will live as one\n\n


In [18]:
beatles[beatles.song=='I\'ll Cry Instead'].text

,text
24774,"I've got every reason on earth to be mad \n'Cause I just lost the only girl I had \nIf I could get my way \nI'd get myself locked up today \nBut I can't, so I'll cry instead \n \nI've got a chip on my shoulder that's bigger that my feet \nI can't talk to people that I meet \nIf I could see you now \nI'd try to make you sad somehow \nBut I can't, so I'll cry instead \n \nDon't want to cry when there's people there \nI get shy when they start to stare \nI'm gonna hide myself away \nBut I'll come back again someday \n \nAnd when you do you'd better hide all the girls \nI'm gonna break their hearts all round the world \nYes, I'm gonna break them in two \nAnd show you what your lovin' man can do \nUntil then I'll cry instead\n\n"


Compare the two lyrics based on the output

**Analysis:** The two songs are not same. The imagine and cry uses words different words. The similarity is they uses emotional words, but their meanings are different.

Find the two Beatles songs that are the most similar
* Using Count Vectorizer
* Using TF-IDF Vectorizer

Calculate the cosine similarity between all combinations of documents

In [19]:

from itertools import combinations

pairs = list(combinations(beatles.song.index, 2)) # all song index combos
pairs_0 = list(combinations(range(len(beatles)), 2)) # all index combos starting with (0,1)
song_pairs = [(beatles.song[a_index], beatles.song[b_index]) for (a_index, b_index) in pairs]
song_pairs[:5]

[('A Shot Of Rhythm And Blues', 'Across The Universe'),
 ('A Shot Of Rhythm And Blues', "All I've Got To Do"),
 ('A Shot Of Rhythm And Blues', 'And I Love Her'),
 ('A Shot Of Rhythm And Blues', 'And Your Bird Can Sing'),
 ('A Shot Of Rhythm And Blues', 'Another Girl')]

 Calculate the cosine similarity of the Count Vectorizer document-term matrix

In [20]:

results = [cosine(X[a_index], X[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, song_pairs), reverse=True)[:5]

[(np.float64(0.8548876342047523), ('All You Need Is Love', 'Love Me Do')),
 (np.float64(0.8359173188630292),
  ('If I Needed Someone', 'If I Needed Someone To Love')),
 (np.float64(0.8058125835571367), ('And I Love Her', 'Love Me Do')),
 (np.float64(0.7930110315846619), ('If I Fell', 'Love Me Do')),
 (np.float64(0.7478391382739419), ('And I Love Her', 'All You Need Is Love'))]

Calculate the cosine similarity of the TF-IDF Vectorizer document-term matrix

In [21]:

results = [cosine(X_tfidf[a_index], X_tfidf[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, song_pairs), reverse=True)[:5]

[(np.float64(0.8717047783261523),
  ('If I Needed Someone', 'If I Needed Someone To Love')),
 (np.float64(0.6599353568343105), ('All You Need Is Love', 'Love Me Do')),
 (np.float64(0.6502842517883831), ('Have A Banana!', 'Crinsk Dee Night')),
 (np.float64(0.6232797327325461), ("Don't Let Me Down", 'Let It Be')),
 (np.float64(0.5275364336322863), ('If I Fell', 'Love Me Do'))]

In [22]:

beatles[beatles.song=='All You Need Is Love'].text

,text
24695,"Love, love, love, love, love, love, love, love, love. \n \nThere's nothing you can do that can't be done. \nNothing you can sing that can't be sung. \nNothing you can say, but you can learn \nHow to play the game \nIt's easy. \nNothing you can make that can't be made. \nNo one you can save that can't be saved. \nNothing you can do, but you can learn \nHow to be you in time \nIt's easy. \n \nAll you need is love, all you need is love, \nAll you need is love, love. Love is all you need. \nLove, love, love, love, love, love, love, love, love. \nAll you need is love, all you need is love, \nAll you need is love, love. Love is all you need. \n \nThere's nothing you can know that isn't known. \nNothing you can see that isn't shown. \nThere's nowhere you can be that isn't where \nYou're meant to be \nIt's easy. \n \nAll you need is love, all you need is love, \nAll you need is love, love. Love is all you need. \nAll you need is love. (All together now). \nAll you need is love. (Everybody). \nAll you need is love, love. Love is all you need. \nLove is all you need. \nLove is all you need \n \n(Yesterday) \n(Oh yeah) \n(She love you, yeah, yeah, yeah) \n(She love you, yeah, yeah, yeah) \n(Oh, yesterday)\n\n"


In [23]:
beatles[beatles.song=='Love Me Do'].text

,text
24805,"Love, love me do. \nYou know I love you, \nI'll always be true, \nSo please, love me do. \nWhoa, love me do. \n \nLove, love me do. \nYou know I love you, \nI'll always be true, \nSo please, love me do. \nWhoa, love me do. \n \nSomeone to love, \nSomebody new. \nSomeone to love, \nSomeone like you. \n \nLove, love me do. \nYou know I love you, \nI'll always be true, \nSo please, love me do. \nWhoa, love me do. \n \nLove, love me do. \nYou know I love you, \nI'll always be true, \nSo please, love me do. \nWhoa, love me do. \nYeah, love me do. \nWhoa, oh, love me do.\n\n"


In [24]:
beatles[beatles.song=='If I Needed Someone'].text


,text
24768,"If I needed someone to love \nYou're the one that I'd be thinking of \nIf I needed someone \n \nIf I had some more time to spend \nThen I guess I'd be with you my friend \nIf I needed someone \nHad you come some other day \nThen it might not have been like this \nBut you see now I'm too much in love \n \nCarve your number on my wall \nAnd maybe you will get a call from me \nIf I needed someone \nAh, ah, ah, ah \n \nIf I had some more time to spend \nThen I guess I'd be with you my friend \nIf I needed someone \nHad you come some other day \nThen it might not have been like this \nBut you see now I'm too much in love \n \nCarve your number on my wall \nAnd maybe you will get a call from me \nIf I needed someone \nAh, ah\n\n"


In [25]:
beatles[beatles.song=='If I Needed Someone To Love'].text

,text
24769,If I needed someone to love \nYoure the woman I'd be thinking of \nIf I needed someone \n \nIf I had some more time to spend \nThen I guess I'd be with you my friend \nIf I needed someone \n \nHad you come some other day \nThen it might not have been like this \nBut you see now I'm too much in love \n \nCarve your number on my wall \nAnd maybe you will get a call from me \nIf I needed someone \n \nIf I had some more time to spend \nThen I guess I'd be with you my friend \nIf I needed someone \n \nHad you come some other day \nThen it might not have been like this \nBut you see now I'm too much in love \n \nCarve your number on my wall \nAnd maybe you will get a call from me \nIf I needed someone \nIf I needed someone to love \nYoure the woman I'd be thinking of \nIf I needed someone \n \nIf I had some more time to spend \nThen I guess I'd be with you my friend \nIf I needed someone \n \nHad you come some other day \nThen it might not have been like this \nBut you see now I'm too much in love \n \nCarve your number on my wall \nAnd maybe you will get a call from me \nIf I needed someone \n \nIf I had some more time to spend \nThen I guess I'd be with you my friend \nIf I needed someone \n \nHad you come some other day \nThen it might not have been like this \nBut you see now I'm too much in love \n \nCarve your number on my wall \nAnd maybe you will get a call from me \nIf I needed someone\n\n


In [26]:
beatles[beatles.song=='Have A Banana!'].text

,text
1221,"[Speech] \n \nBrian Matthew: Is that it? Is that the end? \nPaul: Yeah, yeah, that's it. \nJohn: Fade, fade! \nBrian: Good track. Oh, well, we'll stop there, stop there, stop there. \nJohn: What an end! \nBrian: Quiet! All right, George. \nJohn: Fade! \nBrian: Hold it! \nGeorge: Oh, thank you. \nJohn: Fade, you silly. \nBrian: Well, we did. We did that. Oh, no! No! We've done that bit! \nJohn: The train comes in now. \nBrian: We did that. \nJohn: Yeah. \nBrian: To pove we weren't playing the record, then, you see. 'Cause,\notherwise, there's no point in you being here, is there? Ha, ha, ha! \nJohn: Yeah, we did that, 'cause it sounds just like it, don't it? \nBrian: Pretty cool lot of fellows, aren't you? Here, Ringo, have a banana,\ncatch!\n\n"


In [27]:
beatles[beatles.song=='Crinsk Dee Night'].text

,text
24713,"[Speech] \n \nBrian Matthew: The next few minutes, we're in the lap of the gods and the\nhands of the Beatles. In my young days, when I was a lad, they used to have\nactors in films and now that they-- \nPaul: Yes? \nJohn: Hey! Listen! \nPaul: It's all changed, now, Brian. They're not doing that, no actors. \nJohn: It's all changed, now. \nBrian: But this is what I wonder. In those days, the actors used to say their\nbest bits were left on the cutting room floor. Did you find that? \nJohn: No, no, no, those were the good bits in the film. You should have seen\nthe rest. \nBrian: Yes? \nJohn: Rubbish! \nBrian: Was it, really? \nJohn: Even worse, yes. \nBrian: Who was worst? \nJohn: Oh, Paul. \nBrian: I see. \nPaul: I think John was about the worst. \nJohn: No, it was you. \nPaul: Oh, Ringo was very good. He was. He's a good lad. \nBrian: He was. They're saying he's a new Charlie Chaplin. Do you think that's\nright? \nJohn: He was miming. \nPaul: You, too, with Jason. \nJohn: Oh, yes, he's an old one. Okay, Ring? \nBrian: Beginning to look like that. Now, then-- \nRingo: All right, John. Can you hear me? \nPaul: Can you hear him? Hello! \nBrian: Not really. I hope not. I hope not! \nJohn: We brought you the flowers, Ring. \nBrian: He's not allowed to talk. \nRingo: Eh? \nJohn: We brought you the flowers. \nRingo: Oh, good. \nJohn: And the grapes. \nRingo: Oh, I like grapes. \nPaul: He likes grapes. \nGeorge: We've been to Portugal. \nBrian: Guess who's, guess who's top of the pops in Portugal, then. \nAll: Who? \nBrian: Los Beatles. \nJohn: Los Beatles? Great, great laugh. \nBrian: I don't suppose you know the title of your film in Portuguese? \nJohn: No. Crinsk dee Night? \nBrian: Could be. Let's hear the number, shall we? \nJohn: Right\n\n"


Compare the results. Which Vectorizer seems to do a better job?

**Analysis:** CountVectorizer shows very high similarity scores because songs repeat the word “love”, even if the rest of the lyrics are different. TF‑IDF lowers the importance of those repeated words, so the scores are more balanced and highlight songs that are truly alike.


## Supplementary Activity

* Use a dataset that consists of text document
* Remove special characters
* Compare the similar document using Count Vector and TF-IDF Vectorizer
* Use the cosine similarity to measure the similarity of text


Use a dataset that consists of text document

In [28]:
import pandas as pd

data = pd.read_csv('/content/drive/MyDrive/CPE 313/H.O.A. 8.1/Corona.csv', encoding='latin1')
data = data.drop(columns = ['ScreenName', 'Location', 'TweetAt'])
data.head()

,UserName,OriginalTweet,Sentiment
0,3799,@MeNyrbie @Phil_Gahan @Chrisitv https://t.co/iFz9FAn2Pa and https://t.co/xX6ghGFzCC and https://t.co/I2NlzdxNo8,Neutral
1,3800,advice Talk to your neighbours family to exchange phone numbers create contact list with phone numbers of neighbours schools employer chemist GP set up online shopping accounts if poss adequate supplies of regular meds but not over order,Positive
2,3801,"Coronavirus Australia: Woolworths to give elderly, disabled dedicated shopping hours amid COVID-19 outbreak https://t.co/bInCA9Vp8P",Positive
3,3802,"My food stock is not the only one which is empty...\r\r\n\r\r\nPLEASE, don't panic, THERE WILL BE ENOUGH FOOD FOR EVERYONE if you do not take more than you need. \r\r\nStay calm, stay safe.\r\r\n\r\r\n#COVID19france #COVID_19 #COVID19 #coronavirus #confinement #Confinementotal #ConfinementGeneral https://t.co/zrlG0Z520j",Positive
4,3803,"Me, ready to go at supermarket during the #COVID19 outbreak.\r\r\n\r\r\nNot because I'm paranoid, but because my food stock is litteraly empty. The #coronavirus is a serious thing, but please, don't panic. It causes shortage...\r\r\n\r\r\n#CoronavirusFrance #restezchezvous #StayAtHome #confinement https://t.co/usmuaLq72n",Extremely Negative


Remove special characters

In [29]:
import re
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

def filter_text(text):
    # remove URLs
    text = re.sub(r'http\S+|www\S+', '', text)
    # remove mentions and hashtags
    text = re.sub(r'@\w+|#\w+', '', text)
    # remove special characters and numbers
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    # lowercase
    text = text.lower()
    # strip extra spaces
    text = text.strip()
    # remove stopwords (optional)
    text = ' '.join([word for word in text.split() if word not in stop_words])
    return text

data['clean_tweet'] = data['OriginalTweet'].apply(filter_text)
data = data[data['clean_tweet'].str.strip() != '']
data.head()

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


,UserName,OriginalTweet,Sentiment,clean_tweet
1,3800,advice Talk to your neighbours family to exchange phone numbers create contact list with phone numbers of neighbours schools employer chemist GP set up online shopping accounts if poss adequate supplies of regular meds but not over order,Positive,advice talk neighbours family exchange phone numbers create contact list phone numbers neighbours schools employer chemist gp set online shopping accounts poss adequate supplies regular meds order
2,3801,"Coronavirus Australia: Woolworths to give elderly, disabled dedicated shopping hours amid COVID-19 outbreak https://t.co/bInCA9Vp8P",Positive,coronavirus australia woolworths give elderly disabled dedicated shopping hours amid covid outbreak
3,3802,"My food stock is not the only one which is empty...\r\r\n\r\r\nPLEASE, don't panic, THERE WILL BE ENOUGH FOOD FOR EVERYONE if you do not take more than you need. \r\r\nStay calm, stay safe.\r\r\n\r\r\n#COVID19france #COVID_19 #COVID19 #coronavirus #confinement #Confinementotal #ConfinementGeneral https://t.co/zrlG0Z520j",Positive,food stock one empty please dont panic enough food everyone take need stay calm stay safe
4,3803,"Me, ready to go at supermarket during the #COVID19 outbreak.\r\r\n\r\r\nNot because I'm paranoid, but because my food stock is litteraly empty. The #coronavirus is a serious thing, but please, don't panic. It causes shortage...\r\r\n\r\r\n#CoronavirusFrance #restezchezvous #StayAtHome #confinement https://t.co/usmuaLq72n",Extremely Negative,ready go supermarket outbreak im paranoid food stock litteraly empty serious thing please dont panic causes shortage
5,3804,"As news of the region’s first confirmed COVID-19 case came out of Sullivan County last week, people flocked to area stores to purchase cleaning supplies, hand sanitizer, food, toilet paper and other goods, @Tim_Dodson reports https://t.co/cfXch7a2lU",Positive,news regions first confirmed covid case came sullivan county last week people flocked area stores purchase cleaning supplies hand sanitizer food toilet paper goods reports


In [30]:
data_clean = data[['UserName', 'clean_tweet']].head(100)
data_text = data_clean['clean_tweet']

Compare the similar document using Count Vector and TF-IDF Vectorizer

In [31]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(stop_words='english')
x_cv = cv.fit_transform(data_text).toarray()

x_cv_df = pd.DataFrame(x_cv, columns=cv.get_feature_names_out(), index=data_clean['UserName'])
x_cv_df.head()

,abou,absolutely,according,accounts,accumulation,actions,actively,adara,addition,additional,...,worry,worst,worth,yall,yeah,yes,yesterday,young,youre,zaidon
UserName,,,,,,,,,,,,,,,,,,,,,
3800,0,0,0,1,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3801,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3802,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3803,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3804,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [32]:
from sklearn.feature_extraction.text import TfidfVectorizer

tv = TfidfVectorizer(stop_words='english')
x_tv = tv.fit_transform(data_text).toarray()

x_tv_df = pd.DataFrame(x_tv, columns=tv.get_feature_names_out(), index=data_clean['UserName'])
x_tv_df.head()

,abou,absolutely,according,accounts,accumulation,actions,actively,adara,addition,additional,...,worry,worst,worth,yall,yeah,yes,yesterday,young,youre,zaidon
UserName,,,,,,,,,,,,,,,,,,,,,
3800,0.0,0.0,0.0,0.183948,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3801,0.0,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3802,0.0,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3803,0.0,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3804,0.0,0.0,0.0,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [33]:
from numpy import dot
from numpy.linalg import norm

cosine = lambda v1, v2: dot(v1, v2) / (norm(v1) * norm(v2))

In [34]:
user3800 = list(x_tv_df.iloc[0])
user3800[:20]

[0.0,
 0.0,
 0.0,
 0.18394839122647608,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.18394839122647608,
 0.0,
 0.0,
 0.16879498643551943,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0]

In [35]:
similarity_count = [cosine (user3800, UserName) for UserName in x_cv]
sorted(list(zip(similarity_count, data_clean.UserName)), reverse=True)[1:6]

[(np.float64(0.09635583263759989), 3807),
 (np.float64(0.08483054446957816), 3862),
 (np.float64(0.08240983471677508), 3870),
 (np.float64(0.06671562149471855), 3880),
 (np.float64(0.06390822929019421), 3811)]

In [36]:
similarity_count = [cosine (user3800, UserName) for UserName in x_tv]
sorted(list(zip(similarity_count, data_clean.UserName)), reverse=True)[1:6]

[(np.float64(0.09145248522368905), 3807),
 (np.float64(0.06906697978939276), 3880),
 (np.float64(0.06831144594776724), 3862),
 (np.float64(0.05927378672430961), 3870),
 (np.float64(0.04977305866396603), 3882)]

Use the cosine similarity to measure the similarity of text

In [37]:
from itertools import combinations

pairs = list(combinations(data_clean.UserName.index, 2))
pairs_0 = list(combinations(range(len(data_clean)), 2))
covid_pairs = [(data_clean.UserName[a_index],data_clean.UserName[b_index]) for (a_index, b_index) in pairs]
covid_pairs[:5]

[(np.int64(3800), np.int64(3801)),
 (np.int64(3800), np.int64(3802)),
 (np.int64(3800), np.int64(3803)),
 (np.int64(3800), np.int64(3804)),
 (np.int64(3800), np.int64(3805))]

In [38]:
results = [cosine(x_cv[a_index], x_cv[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, covid_pairs), reverse=True)[:5]

[(np.float64(0.9354143466934852), (np.int64(3821), np.int64(3822))),
 (np.float64(0.44474958999666075), (np.int64(3802), np.int64(3812))),
 (np.float64(0.43301270189221935), (np.int64(3806), np.int64(3834))),
 (np.float64(0.3952847075210474), (np.int64(3850), np.int64(3893))),
 (np.float64(0.3928371006591931), (np.int64(3895), np.int64(3896)))]

In [39]:
results = [cosine(x_tv[a_index], x_tv[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, covid_pairs), reverse=True)[:5]

[(np.float64(0.9212307959748046), (np.int64(3821), np.int64(3822))),
 (np.float64(0.40734814941588876), (np.int64(3850), np.int64(3893))),
 (np.float64(0.35351281215607566), (np.int64(3895), np.int64(3896))),
 (np.float64(0.3524330231913364), (np.int64(3806), np.int64(3834))),
 (np.float64(0.34986164548886856), (np.int64(3802), np.int64(3867)))]

**Analysis:** The results show that Count Vectorizer gives higher similarity scores, especially when tweets share common words. For example, tweets 3821 and 3822 scored 0.935, which is very high. However, this method can be misleading because it treats all repeated words as equally important, even if the tweets have different meanings.
On the other hand, TF-IDF Vectorizer adjusts for word frequency, so common words have less influence. It still found tweets 3821 and 3822 to be very similar (score: 0.921), but other scores were lower and more balanced.


## Conclusion

In this laboratory activity, I learned how to clean and prepare text by removing unnecessary characters, which makes the data easier to analyze. I also applied methods like count vectorizer and tf-idf to convert text into numerical form and used cosine similarity to measure how similar texts are. On the supplementary activity, I understood that while both methods ahve similarities, TF-IDF gives more accurate results because it reduces the common repeated words.